infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java

```diff
@@ -1,24 +1,23 @@
 package org.corfudb.infrastructure;
 
 import io.netty.channel.ChannelHandlerContext;
-import lombok.Getter;
-import lombok.Setter;
 import lombok.extern.slf4j.Slf4j;
 import org.corfudb.protocols.wireprotocol.CorfuMsg;
+import org.corfudb.protocols.wireprotocol.CorfuMsgType;
 
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Created by mwei on 12/4/15.
  */
 @Slf4j
 public abstract class AbstractServer {
 
-    private volatile boolean shutdown;
-
-    public AbstractServer() {
-        shutdown = false;
-    }
+    /**
+     * Current server state
+     */
+    private final AtomicReference<ServerState> state = new AtomicReference<>(ServerState.READY);
```

Comment:
Why you need AtomicReference here? You are not using any atomic operation, volatile would be enough in this case.

Reply:
Because atomic always better that volatile

Reply:
And I use atomic operations, see setState()

Reply:
also https://pmd.github.io/latest/pmd_rules_java_multithreading.html#avoidusingvolatile

Reply:
> Because atomic always better that volatile
Not a valid justification.
https://stackoverflow.com/questions/281132/java-volatile-reference-vs-atomicreference
In case you are using only get and set, volatile is sufficient.

But I see that you are planning to update the state variable atomically if not shutdown.

Reply:
>Not a valid justification.
Your link doesn't explain your point. I still don't see where is my argument is not valid.

Reply:
@xnull First AtomicReference is internally using volatile: https://github.com/bpupadhyaya/openjdk-8/blob/45af329463a45955ea2759b89cb0ebfe40570c3f/jdk/src/share/classes/java/util/concurrent/atomic/AtomicReference.java#L62

Second, your setState() is not atomic, get() and set() are just same as normal operation, unless you need sth. like getAndSet(), compareAndSet() etc.

Reply:
>your setState() is not atomic,

that's not true

Reply:
@xnull why is it atomic?  getState() calls state.get(),  and then you call state.set(), this is not atomic?

Reply:
Now, I got your point. Will fix setState preventing going to incorrect state.
Regarding atomicref - it's just more powerful volatile. Don't see any reason to replace it by volatile

Reply:
Another point is if you need atomic operations like getAndSet(), you need AtomicReference, but if you're only using get() and set(), please use volatile. It will be confusing and pointless if you use AtomicReference, but does not use atomic operations it provides.

Reply:
get and set are not atomic operations? Using Atomic over volatile always better even if you use only get and set

Reply:
@xnull  please check the implementation of get() and set(). They are just doing `value = newValue` and `return value` where value is a volatile. It's just a simple wrapper and exactly same as using volatile.

Reply:
>>your setState() is not atomic

fixed

Reply:
> It's just a simple wrapper and exactly same as using volatile.
Exactly! So
 1. using Atomic over volatile prevents you from accidentally deleting `volatile`. Whenever you use atomic you know exactly that the variable it's atomic.
 2.  Atomic requres you to use it accordingly, with volatile it's much easier to make a mistake.

Reply:
I don't agree with your justification. This is abusing AtomicXXX classes. It is designed to give you ability to use atomic operations, not to replace volatile. Using AtomicReference means you need the atomic operations. Using atomic classes with no atomic operations only creates confusion.

Reply:
It's your opinion) so, up to you. I believe to what I explained